### ISSUE TITLE
`clippy::unnecessary_lazy_evaluations` produces a suggestion that would cause a divide-by-zero panic, despite the original code not causing such

### ISSUE THREAD

**Issue by LikeLakers2** [rust-lang/rust-clippy#11722]
### Summary

As the title says, one of Clippy's lints has produced a suggestion that would cause a divide-by-zero panic. The code and clippy's suggestion are listed below, under the Reproducer header.

### Lint Name

`clippy::unnecessary_lazy_evaluations`

### Reproducer

I tried this code:

```rust
fn clippy_lint_error(num1: usize, num2: usize) -> Option<(usize, usize)> {
	(num2 != 0).then(|| (
		num1 / num2,
		num1 % num2
	))
}
```

I saw this happen:

```
warning: unnecessary closure used with `bool::then`
  --> src\snippet.rs:23:3
   |
23 | /         (num2 != 0).then(|| (
24 | |             num1 / num2,
25 | |             num1 % num2
26 | |         ))
   | |__________^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_lazy_evaluations
   = note: `#[warn(clippy::unnecessary_lazy_evaluations)]` on by default
help: use `then_some(..)` instead
   |
23 ~         (num2 != 0).then_some((
24 ~             num1 / num2,
25 ~             num1 % num2
26 ~         ))
   |
```

I expected to see this happen:

No lint triggering at all - or alternative code that would still avoid the same panic, like:

```rust
// Note: This code does not trigger Clippy
fn clippy_lint_error_fixed(num1: usize, num2: usize) -> Option<(usize, usize)> {
	if num2 != 0 {
		Some((
			num1 / num2,
			num1 % num2
		))
	} else {
		None
	}
}
```

### Version

```text
rustc 1.73.0 (cc66ad468 2023-10-03)
binary: rustc
commit-hash: cc66ad468955717ab92600c770da8c1601a4ff33
commit-date: 2023-10-03
host: x86_64-pc-windows-msvc
release: 1.73.0
LLVM version: 17.0.2
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by LikeLakers2** [comment]
Closing this, since I found an old issue reporting the same thing: #9422
